File: python_code/search/binary_search_recursive.py
"""
Example use:
cat bs_data.csv | python3 binary_search_recursive.py > bsr.csv

"""
import sys
sys.path.append('..')
from util import benchmark_search, test_search


alg = "binary_search"
kind = "recursive"

def binary_search(arr, x):
    return binary_search_rec(arr, 0, len(arr), x)
    
def binary_search_rec(arr, lo, hi, x):
    if (lo == hi):
        return -1
    else:
        mid = (hi + lo) // 2
        if (arr[mid] == x):
            return mid
        elif (arr[mid] > x): 
            return binary_search_rec(arr, lo, mid, x)
        else: # arr[mid] < x 
            return binary_search_rec(arr, mid + 1, hi, x)


if __name__ == '__main__':
    test_search(binary_search)
    benchmark_search(binary_search, alg, kind)

File: c_code/search/binary_search_recursive.c
#include "searchlib.h"

int binary_search_rec(int * arr, int n, int value) {
    return binary_search_rec_helper(arr, 0, n, value);
   }

int binary_search_rec_helper(int * arr, int low, int high, 
    int value) {
    if (low == high) {
        return -1;
    }
    else {
        int mid = (low + high) / 2;
        if (value == arr[mid]) {
            return mid;
        }
        else if (arr[mid] < value) {
            return binary_search_rec_helper(arr, mid + 1, high, value);
        }
        else {
            return binary_search_rec_helper(arr, low, mid, value);
        }
    }
}



